File: examples/saas-test-runner/test_runner.py
#!/usr/bin/env python3
"""Payments API test runner — generates background traffic and error bursts.

Run separately from the app. Uses plain HTTP requests (no SDK import).
Chaos scenarios are triggered via the app's /_ops/simulate/* endpoints.

Usage:
    python test_runner.py

Environment:
    APP_URL  — base URL of the test app (default: http://localhost:8000)
"""

import asyncio
import json
import logging
import os
import random
import sys

from dotenv import load_dotenv

load_dotenv()  # Load .env file if present

import httpx

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(name)s: %(message)s",
)
logger = logging.getLogger("test_runner")

APP_URL = os.getenv("APP_URL", "http://localhost:8000").rstrip("/")

# Weighted endpoint list for traffic generation
ENDPOINTS = [
    ("GET",  "/health",               8),
    ("GET",  "/v1/accounts/acct_1001", 6),
    ("GET",  "/v1/accounts/acct_1002", 4),
    ("POST", "/v1/payments/charge",    5),
    ("POST", "/v1/payments/authorize", 4),
    ("POST", "/v1/transfers/initiate", 3),
    ("GET",  "/v1/rates/convert",      3),
    ("GET",  "/v1/compliance/screen",  2),
    ("POST", "/v1/payments/refund",    2),
]

CURRENCIES = ["USD", "EUR", "GBP"]
MERCHANTS = ["mch_8291", "mch_4517", "mch_7203", "mch_1089", "mch_6345"]
ACCOUNTS = ["acct_1001", "acct_1002", "acct_1003", "acct_1004", "acct_1005"]


def _build_weighted_list() -> list[tuple[str, str]]:
    weighted: list[tuple[str, str]] = []
    for method, path, weight in ENDPOINTS:
        weighted.extend([(method, path)] * weight)
    return weighted


WEIGHTED = _build_weighted_list()


def _make_body(path: str) -> dict | None:
    """Generate a realistic JSON body for POST endpoints."""
    if path == "/v1/payments/charge":
        return {
            "amount": round(random.uniform(15.0, 500.0), 2),
            "currency": random.choice(CURRENCIES),
            "source": f"tok_visa_{random.randint(1000, 9999)}",
            "merchant_id": random.choice(MERCHANTS),
            "idempotency_key": f"idem_{random.randbytes(4).hex()}",
        }
    elif path == "/v1/payments/authorize":
        return {
            "amount": round(random.uniform(5.0, 500.0), 2),
            "currency": random.choice(CURRENCIES),
            "card_last4": str(random.randint(1000, 9999)),
            "merchant_id": random.choice(MERCHANTS),
        }
    elif path == "/v1/transfers/initiate":
        accts = random.sample(ACCOUNTS, 2)
        return {
            "from_account": accts[0],
            "to_account": accts[1],
            "amount": round(random.uniform(50.0, 5000.0), 2),
            "currency": random.choice(CURRENCIES),
        }
    elif path == "/v1/payments/refund":
        return {
            "transaction_id": f"txn_{random.randbytes(3).hex()}",
            "amount": round(random.uniform(10.0, 250.0), 2),
            "reason": random.choice(["customer_request", "duplicate", "fraudulent"]),
        }
    return None


async def traffic_loop() -> None:
    """Send weighted-random requests to the app every ~15s."""
    logger.info("Traffic loop started (target: %s)", APP_URL)
    while True:
        method, path = random.choice(WEIGHTED)
        url = f"{APP_URL}{path}"
        try:
            async with httpx.AsyncClient(timeout=30.0) as client:
                if method == "GET":
                    resp = await client.get(url)
                else:
                    body = _make_body(path)
                    resp = await client.post(url, json=body)
                logger.info("%s %s -> %d", method, path, resp.status_code)
        except Exception as e:
            logger.warning("%s %s -> error: %s", method, path, e)

        delay = random.uniform(10, 20)
        await asyncio.sleep(delay)


async def error_burst_loop() -> None:
    """Send rapid failed authorization/refund requests every ~5 minutes."""
    logger.info("Error burst loop started")
    # Wait a bit before first burst
    await asyncio.sleep(60)

    while True:
        logger.info("=== Starting error burst (20 requests) ===")
        async with httpx.AsyncClient(timeout=30.0) as client:
            for i in range(20):
                try:
                    if i % 3 == 0:
                        body = _make_body("/v1/payments/refund")
                        await client.post(f"{APP_URL}/v1/payments/refund", json=body)
                    else:
                        body = _make_body("/v1/payments/authorize")
                        await client.post(f"{APP_URL}/v1/payments/authorize", json=body)
                except Exception:
                    pass
                await asyncio.sleep(0.2)
        logger.info("=== Error burst complete ===")

        # Wait ~5 minutes before next burst
        delay = random.uniform(270, 330)
        await asyncio.sleep(delay)


async def main() -> None:
    logger.info("=" * 60)
    logger.info("Payments API — Test Runner")
    logger.info("Target app: %s", APP_URL)
    logger.info("=" * 60)

    # Verify app is reachable
    try:
        async with httpx.AsyncClient(timeout=5.0) as client:
            resp = await client.get(f"{APP_URL}/health")
            logger.info("App health check: %d %s", resp.status_code, resp.json())
    except Exception as e:
        logger.error("Cannot reach app at %s: %s", APP_URL, e)
        logger.error("Start the app first: python -m uvicorn app:app --host 0.0.0.0 --port 8000")
        sys.exit(1)

    # Run traffic and error burst loops concurrently
    logger.info("Starting traffic generation...")
    await asyncio.gather(
        traffic_loop(),
        error_burst_loop(),
    )


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        logger.info("Test runner stopped")

File: examples/saas-test-runner-node/test_runner.js
#!/usr/bin/env node
/**
 * Ledger Service test runner — generates traffic and error bursts.
 *
 * Run separately from the app. Uses plain HTTP requests (no SDK import).
 * Chaos scenarios are triggered via the app's /_ops/simulate/* endpoints.
 *
 * Usage:
 *   node test_runner.js
 *
 * Environment:
 *   APP_URL — base URL of the test app (default: http://localhost:8001)
 */

require("dotenv").config();

const APP_URL = (process.env.APP_URL || "http://localhost:8001").replace(/\/+$/, "");

const CURRENCIES = ["USD", "EUR", "GBP"];
const MERCHANT_IDS = ["mch_8291", "mch_4517", "mch_7203", "mch_1089", "mch_6345"];
const ACCOUNTS = ["acct_1001", "acct_1002", "acct_1003", "acct_1004", "acct_1005"];

// Weighted endpoint list for traffic generation
const ENDPOINTS = [
  { method: "GET",  path: "/health",               weight: 8 },
  { method: "GET",  path: "/v1/accounts/acct_1001", weight: 6 },
  { method: "GET",  path: "/v1/accounts/acct_1002", weight: 4 },
  { method: "POST", path: "/v1/payments/charge",    weight: 5 },
  { method: "POST", path: "/v1/payments/authorize", weight: 4 },
  { method: "POST", path: "/v1/transfers/initiate", weight: 3 },
  { method: "GET",  path: "/v1/rates/convert",      weight: 3 },
  { method: "GET",  path: "/v1/compliance/screen",  weight: 2 },
  { method: "POST", path: "/v1/payments/refund",    weight: 2 },
];

function buildWeightedList() {
  const weighted = [];
  for (const { method, path: p, weight } of ENDPOINTS) {
    for (let i = 0; i < weight; i++) weighted.push({ method, path: p });
  }
  return weighted;
}

const WEIGHTED = buildWeightedList();

function log(level, msg, ...args) {
  const ts = new Date().toISOString();
  const formatted = args.length ? `${msg} ${args.join(" ")}` : msg;
  console.log(`${ts} [${level}] test_runner: ${formatted}`);
}

function randomChoice(arr) {
  return arr[Math.floor(Math.random() * arr.length)];
}

function makeBody(path) {
  if (path === "/v1/payments/charge") {
    return {
      amount: parseFloat((Math.random() * 485 + 15).toFixed(2)),
      currency: randomChoice(CURRENCIES),
      source: `tok_visa_${Math.floor(Math.random() * 9000 + 1000)}`,
      merchant_id: randomChoice(MERCHANT_IDS),
      idempotency_key: `idem_${Math.random().toString(36).slice(2, 10)}`,
    };
  }
  if (path === "/v1/payments/authorize") {
    return {
      amount: parseFloat((Math.random() * 495 + 5).toFixed(2)),
      currency: randomChoice(CURRENCIES),
      card_last4: String(Math.floor(Math.random() * 9000 + 1000)),
      merchant_id: randomChoice(MERCHANT_IDS),
    };
  }
  if (path === "/v1/transfers/initiate") {
    const accts = [...ACCOUNTS].sort(() => Math.random() - 0.5).slice(0, 2);
    return {
      from_account: accts[0],
      to_account: accts[1],
      amount: parseFloat((Math.random() * 4950 + 50).toFixed(2)),
      currency: randomChoice(CURRENCIES),
    };
  }
  if (path === "/v1/payments/refund") {
    return {
      transaction_id: `txn_${Math.random().toString(36).slice(2, 8)}`,
      amount: parseFloat((Math.random() * 240 + 10).toFixed(2)),
      reason: randomChoice(["customer_request", "duplicate", "fraudulent"]),
    };
  }
  return null;
}

// -----------------------------------------------------------------------
// Traffic loop — weighted random requests every ~15s
// -----------------------------------------------------------------------

async function trafficLoop() {
  log("INFO", `Traffic loop started (target: ${APP_URL})`);

  while (true) {
    const { method, path: p } = WEIGHTED[Math.floor(Math.random() * WEIGHTED.length)];
    const url = `${APP_URL}${p}`;

    try {
      const opts = { method, signal: AbortSignal.timeout(30000) };
      if (method === "POST") {
        const body = makeBody(p);
        if (body) {
          opts.headers = { "Content-Type": "application/json" };
          opts.body = JSON.stringify(body);
        }
      }
      const resp = await fetch(url, opts);
      log("INFO", `${method} ${p} -> ${resp.status}`);
    } catch (err) {
      log("WARN", `${method} ${p} -> error: ${err.message}`);
    }

    const delay = Math.random() * 10000 + 10000; // 10-20s
    await new Promise((r) => setTimeout(r, delay));
  }
}

// -----------------------------------------------------------------------
// Error burst loop — rapid failed authorizations/refunds every ~5 minutes
// -----------------------------------------------------------------------

async function errorBurstLoop() {
  log("INFO", "Error burst loop started");
  // Wait before first burst
  await new Promise((r) => setTimeout(r, 60000));

  while (true) {
    log("INFO", "=== Starting error burst (20 requests) ===");

    for (let i = 0; i < 20; i++) {
      const isRefund = i % 3 === 0;
      const endpoint = isRefund ? "/v1/payments/refund" : "/v1/payments/authorize";
      const body = makeBody(endpoint);

      try {
        await fetch(`${APP_URL}${endpoint}`, {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify(body),
          signal: AbortSignal.timeout(30000),
        });
      } catch {
        // ignore
      }
      await new Promise((r) => setTimeout(r, 200));
    }

    log("INFO", "=== Error burst complete ===");

    // Wait ~5 minutes before next burst
    const delay = Math.random() * 60000 + 270000; // 270-330s
    await new Promise((r) => setTimeout(r, delay));
  }
}

// -----------------------------------------------------------------------
// Main
// -----------------------------------------------------------------------

async function main() {
  log("INFO", "============================================================");
  log("INFO", "Ledger Service — Test Runner");
  log("INFO", `Target app: ${APP_URL}`);
  log("INFO", "============================================================");

  // Verify app is reachable
  try {
    const resp = await fetch(`${APP_URL}/health`, { signal: AbortSignal.timeout(5000) });
    const data = await resp.json();
    log("INFO", `App health check: ${resp.status} ${JSON.stringify(data)}`);
  } catch (err) {
    log("ERROR", `Cannot reach app at ${APP_URL}: ${err.message}`);
    log("ERROR", "Start the app first: node app.js");
    process.exit(1);
  }

  // Run traffic and error burst loops concurrently
  log("INFO", "Starting traffic generation...");
  await Promise.all([trafficLoop(), errorBurstLoop()]);
}

main().catch((err) => {
  log("ERROR", `Fatal: ${err.message}`);
  process.exit(1);
});
